Add write method

christianbundy opened the pull request:
- Add .gitignore, add .npmrc, update deps 
- Add simple write method to overwrite data 
- Use non-deprecated `Buffer.alloc()` method 

The only backward-incompatible-ish change is that we're no longer throwing errors when we read empty buffers from a file. I'd especially love feedback on the `write()` method in `file.js` where I'm using a new file descriptor -- I couldn't find an elegant way to share the FD because the main one was doing read + append, whereas all we need to do is read + write.

Resolves #4

christianbundy committed:
Add .gitignore, add .npmrc, update deps

christianbundy committed:
Add simple write method to overwrite data

christianbundy committed:
Use non-deprecated `Buffer.alloc()` method

dominictarr commented:
The thing you really need to do is handle updating the blocks cache.
The general approach that is used in append, is append to the file, then update the block.

This is important because reads could come in while you are writing the update.
Hmm, an even simpler way though, is do the write, then make sure that the blocks are evicted from the cache, so that the next read then reloads the block.

dominictarr commented:
I think this could have a edgecase though, because a record can overlap two blocks. you could write to block 1, but mean while, a record that overlaps 1,2 was being read. it reads block 1, then starts to read block 2, but then block 1 is deleted and evicted, and then block 2 is read... hmm actually this might be okay if the read holds a reference to block 1 (I have a feeling that it does...)

dominictarr commented:
confirmed: https://github.com/flumedb/aligned-block-file/blob/master/blocks.js#L60

Btw, I also had this module: https://github.com/dominictarr/flumelog-aligned-offset
that always stores records so that they never overlap the edges of a block.

dominictarr commented on blocks.js:
```diff
@@ -60,17 +60,10 @@ module.exports = function (file, block_size, cache) {
+          next(i+1)
+        } else {
           var buffer = bufs.length == 1 ? bufs[0] : Buffer.concat(bufs)
-          if(!buffer.length)
```
why did you remove this? doesn't your code just return a buffer of zeros? when you read this?

christianbundy replied:
Thanks, nice catch. I thought that this was throwing errors on buffers of nullbytes, but it looks like this still passes tests so I'm not sure where I got that idea. I've re-added it now.

dominictarr commented:
sorry, that was a stream of conciousness.
The test reads the first block, which should load it into the cache, then writes... then reads that again. that should be the correct test... but I don't see anything that clears the cached block. So I don't think that test should pass. prehaps there is an edgecase that makes it work because it's both the first and last block

christianbundy commented on blocks.js:
```diff
@@ -174,6 +166,19 @@ module.exports = function (file, block_size, cache) {
+     */
+    write: (buf, pos, cb) => {
+      const i = Math.floor(pos/block_size)
+      cache.remove(i)
```
> I don't see anything that clears the cached block

@dominictarr did you see this bit? I *think* this clears the cached block, but it's possible I'm doing it wrong.

dominictarr replied:
oh, that looks right! you should do it after the write succeeds though, in case a read gets in and sets that item before you callback. hmm... really reads to that item should be blocked while you are writing...

christianbundy committed:
Re-add empty buffer check

Previously I thought that this was throwing errors on buffers of null
bytes, but it looks like this still passes tests so I'm not sure where I
got that idea.

christianbundy committed:
Clear cache *after* file write is finished

christianbundy commented:
I've pushed a change to ensure that the cache is cleared *after* the file write is finished, but I haven't implemented this bit yet:

> really reads to that item should be blocked while you are writing...

Are there any patterns that jump out to you here? I was thinking of using an observable to watch for a write lock, but the implementations I wrote seemed hacky. Is there a way you'd prefer?

christianbundy committed:
Force reads to wait for positional writes (not appends) to finish

christianbundy commented:
I've added tests for the race condition as well as a check so that positional reads don't happen while a positional write is occurring, but it *does* depend on [this PR](https://github.com/dominictarr/obv/pull/2) to avoid the `rm()` dance.

dominictarr commented:
@christianbundy okay, happy to merge this once my concurns with https://github.com/dominictarr/obv/pull/2 are settled

christianbundy committed:
Use rm() to remove obv listener

christianbundy committed:
Revert to obv@0.0.1 and use conventional `rm()` pattern

christianbundy commented:
Alright, I *think* this is good to go now. I was thinking that the obv changes would be minimal but it sounds like we've got a lot to think about on that front. In the meantime, I've reverted to `obv@0.0.1` and used this pattern:

## Before

```js
obv((value, rm) => {
  if (Math.random() > 0.5) {
    rm()
  }
})
```

## After

```js
let shouldDelete = false
const rm = obv((value) => {
  if (shouldDelete === true) return rm()

  if (Math.random() > 0.5) {
    shouldDelete = true
  }
})
```

dominictarr commented:
Okay that looks good.

dominictarr commented:
I tried to merge this but the tests arn't passing.

```
/home/dominic/c/aligned-block-file/node_modules/obv/index.js:8
      var listener = listeners[i](value)
                                 ^

TypeError: listeners[i] is not a function
    at trigger (/home/dominic/c/aligned-block-file/node_modules/obv/index.js:8:34)
    at Function.many.set (/home/dominic/c/aligned-block-file/node_modules/obv/index.js:33:47)
    at /home/dominic/c/aligned-block-file/file.js:109:21
    at many (/home/dominic/c/aligned-block-file/node_modules/obv/index.js:23:49)
    at offset.once (/home/dominic/c/aligned-block-file/file.js:101:20)
    at Function.many.once (/home/dominic/c/aligned-block-file/node_modules/obv/index.js:39:7)
    at Object.write (/home/dominic/c/aligned-block-file/file.js:92:14)
    at Object.write (/home/dominic/c/aligned-block-file/blocks.js:188:12)
    at /home/dominic/c/aligned-block-file/test/simple.js:112:18
    at file.write (/home/dominic/c/aligned-block-file/blocks.js:190:9)
```

christianbundy commented:
You're right, npm _is_ haunted. I had to `rm -rf node_modules` to repro.

You think that's an obv bug? I'll start looking into it.

dominictarr commented on file.js:
```diff
@@ -65,8 +79,52 @@ module.exports = function (file, block_size, flags) {
+            return
+          } else {
+            writing.set(true)
+            shouldRemove = true
```
these lines should most probably be the other way around.
firstly, calling `writing.set(true)` can trigger listeners, which could call other functions, so basically anything could happen. Then you are setting a state that changes how the thing behaves. In some cases, calling `writing.set()` will interact with `shouldRemove` and having being bitten by this, it's very hard to debug things like this!

Here is a case where I got lucky and figured it out quickly: https://github.com/dominictarr/stream-spec/issues/4

Also the [infamous DAO hack](https://coincentral.com/blockchain-security-vulnerabilities/) also depended on a pattern like this.

christianbundy replied:
Hmm, I had it that way originally but it was complaining:

```
/home/christianbundy/src/aligned-block-file/file.js:103
            return rm()
            ^

ReferenceError: rm is not defined
    at Array.<anonymous> (/home/christianbundy/src/aligned-block-file/file.js:103:13)
    at trigger (/home/christianbundy/src/aligned-block-file/node_modules/obv/index.js:8:34)
    at Function.many.set (/home/christianbundy/src/aligned-block-file/node_modules/obv/index.js:33:47)
    at /home/christianbundy/src/aligned-block-file/file.js:110:21
    at many (/home/christianbundy/src/aligned-block-file/node_modules/obv/index.js:23:49)
    at offset.once (/home/christianbundy/src/aligned-block-file/file.js:101:20)
    at Function.many.once (/home/christianbundy/src/aligned-block-file/node_modules/obv/index.js:39:7)
    at Object.write (/home/christianbundy/src/aligned-block-file/file.js:92:14)
    at Object.write (/home/christianbundy/src/aligned-block-file/blocks.js:188:12)
    at /home/christianbundy/src/aligned-block-file/test/simple.js:102:12
```

Even stranger, checking `typeof rm` was throwing the same issue. The only way I can get it working is to replace `rm()` with `setImmediate(() => rm())`.

dominictarr replied:
my hunch is it's the scoping rules of `const` with `var rm` it would work I bet.

dominictarr commented:
hmm, added some quick console.logs and it seems to removing a listener while iterating the listeners.
funny because I thought I fixed this... but I guess we were just never hitting it.

christianbundy committed:
Refactor write/remove order and use `setImmediate(() => rm())`
